usaspending_api/broker/management/commands/load_fpds_transactions.py

```diff
@@ -0,0 +1,158 @@
+from django.core.management.base import BaseCommand
+import logging
+import re
+import psycopg2
+from datetime import datetime, timezone
+
+from usaspending_api.data_load.fpds_loader import load_ids, destroy_orphans, delete_stale_fpds
+from usaspending_api.common.retrieve_file_from_uri import RetrieveFileFromUri
+from usaspending_api.common.helpers.date_helper import datetime_command_line_argument_type
+from usaspending_api.common.helpers.sql_helpers import get_broker_dsn_string
+from usaspending_api.common.helpers.etl_helpers import update_c_to_d_linkages
+from usaspending_api.etl.award_helpers import update_awards, update_contract_awards
+from usaspending_api.broker.helpers.last_load_date import get_last_load_date, update_last_load_date
+
+logger = logging.getLogger("console")
+
+BROKER_CONNECTION_STRING = get_broker_dsn_string()
+
+CHUNK_SIZE = 5000
+
+ALL_FPDS_QUERY = "SELECT {} FROM detached_award_procurement"
+
+
+class Command(BaseCommand):
+    help = "Sync USAspending DB FPDS data using Broker for new or modified records and S3 for deleted IDs"
+
+    modified_award_ids = []
+
+    @staticmethod
+    def get_cursor_for_date_query(connection, date, count=False):
+        if count:
+            db_cursor = connection.cursor()
+            db_query = ALL_FPDS_QUERY.format("COUNT(*)")
+        else:
+            db_cursor = connection.cursor("fpds_load", cursor_factory=psycopg2.extras.DictCursor)
+            db_query = ALL_FPDS_QUERY.format("detached_award_procurement_id")
+
+        if date:
+            db_cursor.execute(db_query + " WHERE updated_at >= %s", [date])
+        else:
+            db_cursor.execute(db_query)
+        return db_cursor
+
+    def load_fpds_from_date(self, date):
+        if date is None:
+            logger.info("fetching all fpds transactions...")
+        else:
+            logger.info("fetching fpds transactions since {}...".format(str(date)))
+
+        stale_awards = delete_stale_fpds(date)
```

Comment:
Deletes don't make sense when we're running a reload since those deleted IDs no longer exist in the source Broker DB.
The workflow is for "incremental" loads (`--since-last-load`). Up to you if you want slightly more complex logic/state to track "incremental" runs or require a flag to be provided when the operator wishes deletes from S3 to occur.

Reply:
We may need to talk offline about this delete requirement. The old code always deletes, but only deletes what it's about to add in the event specific ids are listed to load. If the desired behavior is something other than that, I'll need to be informed.

Reply:
Sure, after standup we can huddle at a desk.
Old loader was only intended for incremental loads, this loader is more capable.

Reply:
After a brief chat with Alisa and some thoughts, here's how I see it:

- if the transaction in in broker's table, it is always valid to load
- an incremental loader (--date or --since-last-load) will always leave everything in the correct state from that date assuming it deletes first and then loads, which it does now
- a spot loader (--file or --ids) will load and NOT delete, which may leave behind transactions that should have been deleted but weren't, but will NOT leave behind transactions that shouldn't have been deleted but were. It cannot, however, "bring back" a transaction that should be deleted and was, because then it wouldn't have been in broker's table
- if a spot loader is run before an incremental loader, the latter will cover the deletes of the former. If an incremental loader is run before a spot loader, the latter will not be able to load anything the incremental loader left in a deleted state
- full-reload is effectively an incremental loader, which the exception that it doesn't have a date with which to find S3 files

Therefore, I think that the only change that needs to be made is to put in a date for the delete in the event of a full reload.

Reply:
I think the comment is accurate, there is some confusion on my end between deletes being sourced from S3 file and if deletes occur on "spot" loads using the provided IDs.

- an incremental loader (--date or --since-last-load) will always leave everything in the correct state from that date assuming it deletes first and then loads, which it does now
> "Deletes *source from S3* first"

- a spot loader (--file or --ids) will load and NOT delete, 
Not using S3 deletes makes sense. Should the transactions be purged before being loaded if they already exist in the USAspending DB?

Reply:
I don't see much point; if values exist they will be overwritten.

Reply:
What if the transactions don't exist in Broker?

Reply:
If a spot loader loads a transaction that was removed from broker, it won't update that transaction. That does leave us in an inconsistent state, but if the record was marked for deletion (I'm assuming S3 is the source of truth for that) it will still be removed by the next incremental load, assuming it wasn't removed by an incremental load already. The alternatives would be to have the spot loader delete all transactions it's about to load, search S3 for all time, or add logic to do a delete instead of an upsert when the pull from broker fails for a transaction. 
  The first two have performance concerns, while the last one introduces a second source of truth. I'd argue that that third option would still be the best, but it would add a bit of development time. The current implementation still promises eventual consistency, if not immediate consistency.

Reply:
We'll leave it as-is for now. The "eventual consistency" is based on a number of assumptions. Thankfully we don't run spot load jobs on production very often.